### PR TITLE
new(CI): add ppc64le support to old bpf and kmod plus CI job

### PR DESCRIPTION
**Pull request by mdafsanhossain**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area build

/area CI

/area driver-kmod

/area driver-bpf

/area driver-modern-bpf

/area libsinsp

/area tests

**What this PR does / why we need it**:

This PR adds GHA job to build `kmod,bpf,modern-bpf` on ppc64le (using QEMU) and also run libsinsp unit tests. It is similar to what is currently enabled for `s390x` [here](https://github.com/falcosecurity/libs/blob/4e9a3cc18e40395d3c1d0c73485dd308ae3f875a/.github/workflows/ci.yml#L326).

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: driver/bpf/CMakeLists.txt
Comment (FedeDP) [inline driver/bpf/CMakeLists.txt:18]: I am pretty sure that eBPF support will also need same fixes as modern eBPF, ie: 
* https://github.com/falcosecurity/libs/blob/master/driver/bpf/plumbing_helpers.h#L125
* https://github.com/falcosecurity/libs/blob/master/driver/bpf/plumbing_helpers.h#L167
* https://github.com/falcosecurity/libs/blob/master/driver/bpf/plumbing_helpers.h#L286

I also noticed that we skipped an include here: https://github.com/falcosecurity/libs/blob/master/userspace/libscap/engine/gvisor/parsers.cpp#L35.
I am not sure whether it is needed, but i guess so given we added aarch64 and s390x too (btw gvisor should only be supported on x86_64 atm, since the option is only available on it: https://github.com/falcosecurity/libs/blob/master/cmake/modules/engine_config.cmake#L26)

File: driver/CMakeLists.txt
Comment (FedeDP) [inline driver/CMakeLists.txt:30]: Like for bpf, kmod too needs some small work:
* https://github.com/falcosecurity/libs/blob/master/driver/main.c#L2144
* https://github.com/falcosecurity/libs/blob/master/driver/ppm.h#L81

File: .github/workflows/ci.yml
Comment (FedeDP) [inline .github/workflows/ci.yml:365]: This is a nice idea! Can we use a `matrix` to merge together this job with the build-libs-s390x one?

File: driver/bpf/CMakeLists.txt
Comment (mdafsanhossain) [reply driver/bpf/CMakeLists.txt:18]: Okay. I was able to build bpf and kmod out of the box without any changes, so I assumed I did not need to make any other modifications. I will add the fixes you suggested.

File: .github/workflows/ci.yml
Comment (mdafsanhossain) [reply .github/workflows/ci.yml:365]: Yes we can. I will try it out.

Comment (Andreagit97) [reply .github/workflows/ci.yml:365]: this is a great point, moreover, I just thought that the s390x job takes almost 1 h to run but in the end, we want to check the build of our drivers (modern_bpf, kmod, bpf) why do we need to build `sinsp` and run tests? In order to not slow down our CI too much i would propose to just build the drivers in this CI job, and maybe we can do the same for the s390x WDYT?

Comment (mdafsanhossain) [reply .github/workflows/ci.yml:365]: Do you recommend a separate job for testing `sinsp` or do we stick to the driver builds for now?

Comment (FedeDP) [reply .github/workflows/ci.yml:365]: I think sticking with the drvers build for now is fine; higher level code rarely needs adjustments for new architecture, therefore we are pretty confident that we cannot break the build of scap/sinsp for a given architecture.

Comment (Andreagit97) [reply .github/workflows/ci.yml:365]: yep let's say that breaking the build of the userspace part should never happen and a 1h job at every build seems a high price to pay

Comment (mdafsanhossain) [reply .github/workflows/ci.yml:365]: Makes sense. I will modify for both archs. @hbrueckner let me know if this is ok for `s390x`

Comment (hbrueckner) [reply .github/workflows/ci.yml:365]: HI @mdafsanhossain 
> Makes sense. I will modify for both archs. @hbrueckner let me know if this is ok for `s390x`

@Andreagit97 @FedeDP let's reduce the build time on `s390x` by focusing on the drivers only.

File: driver/CMakeLists.txt
Comment (mdafsanhossain) [reply driver/CMakeLists.txt:30]: @FedeDP I was going through this https://github.com/falcosecurity/libs/blob/c2fd3086ff1bd0fce20a24a6da9f03c47332d68a/driver/feature_gates.h#L64
We need this patch for ppc64 too?

Comment (FedeDP) [reply driver/CMakeLists.txt:30]: I am not sure! That depends if `ppc64le` is hit by the kernel behavior of not sending the clone exit event for the child; at the moment, all supported architectures but x86_64 are hit.
We have drivers tests for it though, eg: https://github.com/falcosecurity/libs/blob/master/test/drivers/test_suites/syscall_exit_suite/clone3_x.cpp#L191
So, if this test is passing fine running on ppc64le, we should not need the CAPTURE_SCHED_PROC_FORK enabled.
The same goes for CAPTURE_SCHED_PROC_EXEC.

cc @Andreagit97

Comment (Andreagit97) [reply driver/CMakeLists.txt:30]: uhm it seems like `ppc64le` behaves like `x86` since looking at failed tests here https://github.com/falcosecurity/libs/pull/1475#issuecomment-1803296626 `execve` and `clone` test are passing. 

BTW always looking at the tests there is another issue:
```
[  FAILED  ] SyscallExit.execveatX_correct_exit
[  FAILED  ] SyscallExit.execveatX_execve_exit
[  FAILED  ] SyscallExit.execveatX_success_memfd
```
it seems that `ppc64le` correctly returns an `execveat` exit event when an `execveat` syscall is called (like s390x and riscv)

https://github.com/falcosecurity/libs/blob/c2fd3086ff1bd0fce20a24a6da9f03c47332d68a/test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp#L214

Other architectures like x86 instead return an `execve` instead of an `execveat` exit event.

Comment (FedeDP) [reply driver/CMakeLists.txt:30]: > but should that be done with a separate PR?

We can do it in this one, since it is implementing support for kmod and old ebpf too :)

File: .github/workflows/ci.yml
Comment (FedeDP) [reply .github/workflows/ci.yml:365]: Also please make sure to share `drivers_test` results on kmod,ebpf and modern ebpf once this is ready, thank you!

File: driver/CMakeLists.txt
Comment (FedeDP) [reply driver/CMakeLists.txt:30]: Also please make sure to share drivers_test results on kmod,ebpf and modern ebpf once this is ready, thank you!

Comment (mdafsanhossain) [reply driver/CMakeLists.txt:30]: Sure, will do.

Comment (mdafsanhossain) [reply driver/CMakeLists.txt:30]: I am trying to run drivers_test for ebpf probe but I am getting this problem

```
-----------------------------------------------------
- Setup phase
-----------------------------------------------------

* Configure BPF probe tests! Probe path: /root/afsan/libs/build/driver/bpf/probe.o
* Using buffer dim: 8388608
-- BEGIN PROG LOAD LOG --
insn 6910 cannot be patched due to 16-bit range
processed 40396 insns (limit 1000000) max_states_per_insn 1 total_states 4057 peak_states 4057 mark_read 73

-- END PROG LOAD LOG --
Unable to open the engine: libscap: bpf_load_program() event=raw_tracepoint/filler/sys_procexit_e: Operation not permitted

```
Running this on a RHEL 9.2 machine.
Kernel version: `5.14.0-284.25.1.el9_2.ppc64le`

Steps I have followed:

1. `cmake -DBUILD_BPF=ON -DBPF_DEBUG=ON -DUSE_BUNDLED_DEPS=On -DENABLE_DRIVERS_TESTS=On -DBUILD_LIBSCAP_GVISOR=Off -DCREATE_TEST_TARGETS=On ..`
2. `sudo ./test/drivers/drivers_test -b`

Kmod and modern-bpf tests work fine.

Comment (FedeDP) [reply driver/CMakeLists.txt:30]: That's interesting; it is an ebpf verifier failure that we never encountered; the issue comes from https://android.googlesource.com/kernel/common/+/4e2c7b297431/kernel/bpf/verifier.c#10490 but we will need to investigate further what's happening.

Comment (FedeDP) [reply driver/CMakeLists.txt:30]: Btw i think you can open a bug report about the issue, but we can go on with this.

Comment (mdafsanhossain) [reply driver/CMakeLists.txt:30]: Yes. I was digging around this but no luck so far. https://github.com/falcosecurity/libs/blob/041477e3acbf3ab79799bfbf28c259fde8ae7e52/driver/bpf/fillers.h#L5203
I will create a bug report for now.

Comment (FedeDP) [reply driver/CMakeLists.txt:30]: If you want to give a shot, you can try to comment the 
```
struct list_head *head = &(task->children);
	struct list_head *next_child = (struct list_head *)_READ(head->next);
	if(next_child != head)
	{
		/* We have at least one child, so we need a reaper for it */
		reaper_pid = find_new_reaper_pid(data, task);
	}
``` 
lines and check if it works!

Comment (mdafsanhossain) [reply driver/CMakeLists.txt:30]: Tests are running now. Is `*next_child` pointing to NULL?

Comment (mdafsanhossain) [reply driver/CMakeLists.txt:30]: Logs for all three drivers test.

[kmod_tests.log](https://github.com/falcosecurity/libs/files/13486720/kmod_tests.log)
[core_bpf_tests.log](https://github.com/falcosecurity/libs/files/13486721/core_bpf_tests.log)
[bpf_tests.log](https://github.com/falcosecurity/libs/files/13486722/bpf_tests.log)

Comment (FedeDP) [reply driver/CMakeLists.txt:30]: > Tests are running now. Is *next_child pointing to NULL?

@Andreagit97 any idea?
@mdafsanhossain can you try uncommenting that piece of code and commenting only the `reaper_pid = find_new_reaper_pid(data, task);` line too?

Comment (mdafsanhossain) [reply driver/CMakeLists.txt:30]: Tried it. That seems to be the offending line

Comment (Andreagit97) [reply driver/CMakeLists.txt:30]: Uhm it seems strange

```c
	struct list_head *head = &(task->children);
	struct list_head *next_child = (struct list_head *)_READ(head->next);
	if(next_child != head)
	{
		/* We have at least one child, so we need a reaper for it */
		reaper_pid = find_new_reaper_pid(data, task);
	}
```

if `head!= null`, `next_child` cannot be `null` because in the worst case, it would be `next_child==head`, BTW we never dereference `next_child` so it shouldn't be an issue at all :thinking:

Comment (FedeDP) [reply driver/CMakeLists.txt:30]: Yep the issue is within `find_new_reaper_pid` :/

**Comment by FedeDP** [comment]
Seems to work reliably; added the PR: https://github.com/falcosecurity/libs/pull/1517

**Comment by FedeDP** [comment]
So, what we need here is the CI part, right? Ie: the matrix merging s390x and ppc64le (and stopping at drivers build without testing libsinsp unit tests).

**Comment by mdafsanhossain** [comment]
> So, what we need here is the CI part, right? Ie: the matrix merging s390x and ppc64le (and stopping at drivers build without testing libsinsp unit tests).

Yes. I am testing it. I will commit in some time

**Comment by FedeDP** [comment]
> Yes. I am testing it. I will commit in some time

No problem, it was just to recap :) take your time! And thank you!

File: driver/CMakeLists.txt
Comment (FedeDP) [reply driver/CMakeLists.txt:30]: I opened the issue to track the legacy bpf verifier failure: https://github.com/falcosecurity/libs/issues/1521

**Comment by mdafsanhossain** [comment]
@FedeDP Can you review my CI changes?

**Comment by FedeDP** [comment]
> @FedeDP Can you review my CI changes?

Sure thing!

File: test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp
Comment (FedeDP) [inline test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: In this file you are using `__powerpc__` in place of `__ppc64le__`. Is it fine?

Comment (FedeDP) [reply test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: See userspace/libscap/engine/gvisor/parsers.cpp changes.

Comment (mdafsanhossain) [reply test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: Yes. `__ppc64le__` does not work.

Comment (FedeDP) [reply test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: So we need to use `__powerpc__` in userspace/libscap/engine/gvisor/parsers.cpp ?

Comment (FedeDP) [reply test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: What about `__powerpc64__` (looking at this: https://reviews.llvm.org/D137513)?

Comment (mdafsanhossain) [reply test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: RIght. I will update it. Honestly, it is a bit confusing with different nomenclatures.

Comment (FedeDP) [reply test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: Agree...

File: userspace/libscap/engine/gvisor/parsers.cpp
Comment (FedeDP) [inline userspace/libscap/engine/gvisor/parsers.cpp:36]: While you are here, can you add the 
```
#elif __riscv
#include "../../driver/syscall_compat_riscv64.h"
#endif
```

Comment (mdafsanhossain) [reply userspace/libscap/engine/gvisor/parsers.cpp:36]: Will do!

File: test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp
Comment (mdafsanhossain) [reply test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: Alternatively, I can use `__PPC64__`
https://www.ibm.com/docs/vi/xl-c-and-cpp-linux/16.1.0?topic=macros-related-platform

Comment (FedeDP) [reply test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: I'd go with the lowercase one if that works :D

Comment (mdafsanhossain) [reply test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp:214]: That works. :D. I guess `__powerpc64__` makes more sense.

File: .github/workflows/ci.yml
Comment (FedeDP) [inline .github/workflows/ci.yml:365]: TO avoid issues like: https://github.com/falcosecurity/libs/pull/1522, i'd build `drivers_test` executable instead of `scap-open`.
TO enable its build, add `-DENABLE_DRIVERS_TESTS=On-DCREATE_TEST_TARGETS=On ` to the cmake configure step!

Comment (FedeDP) [reply .github/workflows/ci.yml:365]: Everything else LGTM! Thank you very much!
We are approaching the end :)

Comment (mdafsanhossain) [reply .github/workflows/ci.yml]: Thanks a ton for the review and help with the patches. :)

**Comment by FedeDP** [comment]
There is a small conflict to be solved in ci.yml (pretty simple, `master` removed some whitespaces!)

File: driver/bpf/CMakeLists.txt
Comment (mdafsanhossain) [reply driver/bpf/CMakeLists.txt:18]: @FedeDP Btw, I noticed these 2 lines point to incorrect links for each architecture.
https://github.com/falcosecurity/libs/blob/f61e7fff24f951af1bbe033416f09fdd439ab6b0/driver/bpf/plumbing_helpers.h#L268

https://github.com/falcosecurity/libs/blob/f61e7fff24f951af1bbe033416f09fdd439ab6b0/driver/bpf/plumbing_helpers.h#L289

Comment (FedeDP) [reply driver/bpf/CMakeLists.txt:18]: I think `libbpf` touched those files (we should've linked a fixed hash instead of master). No problem anyway!

**Review by FedeDP — APPROVED** [review]
/approve

**Review by Andreagit97 — APPROVED** [review]
unfortunately, the build is really slow also without the sinsp tests :/ BTW we can think of something

/approve

**Comment by Andreagit97** [comment]
waiting for the CI to complete

**Comment by mdafsanhossain** [comment]
Let me add `fail-fast: false` as the job strategy before merging

**Comment by FedeDP** [comment]
While you are here @mdafsanhossain care to update naming of the jobs from "build-libs-" to `build-drivers-`

**Review by FedeDP — APPROVED** [review]
/approve

**Review by Andreagit97 — APPROVED** [review]
/approve

**Pull request merged**
